packages/labs/src/components/select/select.tsx

```diff
@@ -291,10 +310,31 @@ export class Select<T> extends React.Component<ISelectProps<T>, ISelectState<T>>
     };
 
     private handleQueryChange = (event: React.FormEvent<HTMLInputElement>) => {
-        const { inputProps = {} } = this.props;
-        this.setState({ query: event.currentTarget.value });
+        const { inputProps = {}, onQueryChange } = this.props;
+        const query = event.currentTarget.value;
+        // Only set the query state ourselves if not controlled
+        if (getQuery(this.props) === undefined) {
+            this.setState({ query });
+        }
         Utils.safeInvoke(inputProps.onChange, event);
+        Utils.safeInvoke(onQueryChange, query);
     };
 
-    private resetQuery = () => this.setState({ activeItem: this.props.items[0], query: "" });
+    private resetQuery = () => {
+        const { items, onQueryChange } = this.props;
+        const query = "";
+        this.setState({ activeItem: items[0], query });
```

Comment:
Seems weird to let the query be mutated in controlled mode. Shouldn't we just emit the _desired_ new query via `onQueryChange`, and then let the caller change the controlled value? Otherwise we get into a weird hybrid place.

Reply:
might be useful to introduce a helper function that can be used by both `resetQuery` and `handleQueryChange` to ensure the correct events are called.

Reply:
Great catch, thanks!